mlflow/pyfunc/__init__.py

```diff
@@ -1633,6 +1644,10 @@ def batch_predict_fn(pdf, params=None):
             if scoring_server_proc is not None:
                 os.kill(scoring_server_proc.pid, signal.SIGTERM)
 
+        # restore original environment variables
+        for k, v in original_envs.items():
+            os.environ.pop(k) if v is None else os.environ.update({k: v})
```

Comment:
shouldn't we put this in a "finally" block ? for restoring when exeception happens.

Reply:
Seemingly the restoring logic has some issue when the "extra_env" dict has overlapped keys with existing env dict.

I recommend we add a helper method like:

```
import contextlib
import os


@contextlib.contextmanager
def modified_environ(*remove, **update):
    """
    Temporarily updates the ``os.environ`` dictionary in-place.

    The ``os.environ`` dictionary is updated in-place so that the modification
    is sure to work in all situations.

    :param remove: Environment variables to remove.
    :param update: Dictionary of environment variables and values to add/update.
    """
    env = os.environ
    update = update or {}
    remove = remove or []

    # List of environment variables being updated or removed.
    stomped = (set(update.keys()) | set(remove)) & set(env.keys())
    # Environment variables and values to restore on exit.
    update_after = {k: env[k] for k in stomped}
    # Environment variables and values to remove on exit.
    remove_after = frozenset(k for k in update if k not in env)

    try:
        env.update(update)
        [env.pop(k, None) for k in remove]
        yield
    finally:
        env.update(update_after)
        [env.pop(k) for k in remove_after]
```
to implement the accurate restoration.

Reply:
This makes sense!

Reply:
But what's the problem with existing implementation? I think we pop the key if it's newly set and restore the value if it exists before.

Reply:
> I think we pop the key if it's newly set and restore the value if it exists before.

double checked, you are right. but I think move the logic to a helper function is more clear, and helper function easier to reuse.